lib/blockstore/badger/blockstore.go

```diff
@@ -0,0 +1,416 @@
+package badgerbs
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync/atomic"
+
+	"github.com/dgraph-io/badger/v2"
+	"github.com/dgraph-io/badger/v2/options"
+	"github.com/multiformats/go-base32"
+
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	logger "github.com/ipfs/go-log/v2"
+	pool "github.com/libp2p/go-buffer-pool"
+
+	"github.com/filecoin-project/lotus/lib/blockstore"
+)
+
+var (
+	// KeyPool is the buffer pool we use to compute storage keys.
+	KeyPool *pool.BufferPool = pool.GlobalPool
+)
+
+var (
+	// ErrBlockstoreClosed is returned from blockstore operations after
+	// the blockstore has been closed.
+	ErrBlockstoreClosed = fmt.Errorf("badger blockstore closed")
+
+	log = logger.Logger("badgerbs")
+)
+
+// aliases to mask badger dependencies.
+const (
+	// FileIO is equivalent to badger/options.FileIO.
+	FileIO = options.FileIO
+	// MemoryMap is equivalent to badger/options.MemoryMap.
+	MemoryMap = options.MemoryMap
+	// LoadToRAM is equivalent to badger/options.LoadToRAM.
+	LoadToRAM = options.LoadToRAM
+)
+
+// Options embeds the badger options themselves, and augments them with
+// blockstore-specific options.
+type Options struct {
+	badger.Options
+
+	// Prefix is an optional prefix to prepend to keys. Default: "".
+	Prefix string
+}
+
+func DefaultOptions(path string) Options {
+	return Options{
+		Options: badger.DefaultOptions(path),
+		Prefix:  "",
+	}
+}
+
+// badgerLog is a local wrapper for go-log to make the interface
+// compatible with badger.Logger (namely, aliasing Warnf to Warningf)
+type badgerLog struct {
+	logger.ZapEventLogger
+}
+
+func (b *badgerLog) Warningf(format string, args ...interface{}) {
+	b.Warnf(format, args...)
+}
+
+const (
+	stateOpen int64 = iota
+	stateClosing
+	stateClosed
+)
+
+// Blockstore is a badger-backed IPLD blockstore.
+//
+// NOTE: once Close() is called, methods will try their best to return
+// ErrBlockstoreClosed. This will guaranteed to happen for all subsequent
+// operation calls after Close() has returned, but it may not happen for
+// operations in progress. Those are likely to fail with a different error.
+type Blockstore struct {
+	DB *badger.DB
+
+	// state is guarded by atomic.
+	state int64
+
+	prefixing bool
+	prefix    []byte
+	prefixLen int
+}
+
+var _ blockstore.Blockstore = (*Blockstore)(nil)
+var _ blockstore.Viewer = (*Blockstore)(nil)
+var _ io.Closer = (*Blockstore)(nil)
+
+// Open creates a new badger-backed blockstore, with the supplied options.
+func Open(opts Options) (*Blockstore, error) {
+	opts.Logger = &badgerLog{*log}
```

Comment:
To get line infromation the badgerLog needs to be a bit more complex:
See: https://github.com/ipfs/go-ds-badger2/blob/master/compat_logger.go
and: https://github.com/ipfs/go-ds-badger2/blob/master/datastore.go#L145-L148

Reply:
Fixing the log would be nice but we can defer it.

Reply:
Missed this. Looking!

Reply:
Addressed in bc6965c.